File: elixir/2342_maximum_sum.exs
# https://leetcode.com/problems/max-sum-of-a-pair-with-equal-sum-of-digits/
# 12/02/2025

# Map wins!
defmodule Solution do
  @spec maximum_sum(nums :: [integer]) :: integer
  def maximum_sum(nums), do: maximum_sum(nums, %{}, -1)

  defp maximum_sum([], _sum_map, max_sum), do: max_sum

  defp maximum_sum([num | nums_rest], sum_map, max_sum) do
    sum_digit = sum_digits(num)
    n = Map.get(sum_map, sum_digit, 0)

    new_max_sum = if n > 0, do: max(max_sum, num + n), else: max_sum
    new_sum_map = Map.put(sum_map, sum_digit, max(n, num))

    maximum_sum(nums_rest, new_sum_map, new_max_sum)
  end

  defp sum_digits(num) do
    Integer.digits(num)
    |> Enum.sum()
  end
end

# Erlang array
# defmodule Solution do
#   @spec maximum_sum(nums :: [integer]) :: integer
#   def maximum_sum(nums), do: maximum_sum(nums, :array.new(82, default: 0), -1)

#   defp maximum_sum([], _sum_arr, max_sum), do: max_sum

#   defp maximum_sum([num | nums_rest], sum_arr, max_sum) do
#     sum_digit = sum_digits(num)
#     n = :array.get(sum_digit, sum_arr)

#     new_max_sum = if n > 0, do: max(max_sum, num + n), else: max_sum
#     new_sum_arr = :array.set(sum_digit, max(n, num), sum_arr)

#     maximum_sum(nums_rest, new_sum_arr, new_max_sum)
#   end

#   defp sum_digits(num) do
#     Integer.digits(num)
#     |> Enum.sum()
#   end
# end

# Tuple
# defmodule Solution do
#   @spec maximum_sum(nums :: [integer]) :: integer
#   def maximum_sum(nums), do: maximum_sum(nums, Tuple.duplicate(0, 81), -1)

#   defp maximum_sum([], _sum_tup, max_sum), do: max_sum

#   defp maximum_sum([num | nums_rest], sum_tup, max_sum) do
#     sum_digit = sum_digits(num)
#     n = elem(sum_tup, sum_digit - 1)

#     new_max_sum = if n > 0, do: max(max_sum, num + n), else: max_sum
#     new_sum_tup = put_elem(sum_tup, sum_digit - 1, max(n, num))

#     maximum_sum(nums_rest, new_sum_tup, new_max_sum)
#   end

#   defp sum_digits(num) do
#     num
#     |> Integer.digits()
#     |> Enum.sum()
#   end
# end

defmodule Main do
  def main do
    test_cases = [
      {
        [18, 43, 36, 13, 7],
        54
      },
      {
        [10, 12, 19, 14],
        -1
      }
    ]

    test_cases
    |> Enum.each(fn {nums, expected} ->
      result = Solution.maximum_sum(nums)
      IO.puts("Input: #{inspect(nums)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Result: #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/2342_maximum_sum.rs
// https://leetcode.com/problems/max-sum-of-a-pair-with-equal-sum-of-digits/
// 12/02/2025

impl Solution {
    pub fn maximum_sum(nums: Vec<i32>) -> i32 {
        let mut sum_arr = [0; 81];
        let mut result = -1;

        for num in nums {
            let sum_digit = Self::sum_digits(num) - 1;
            let n = sum_arr[sum_digit];

            if n > 0 {
                result = result.max(num + n);
            }

            sum_arr[sum_digit] = n.max(num);
        }

        result
    }

    #[inline(always)]
    fn sum_digits(num: i32) -> usize {
        let mut num = num;
        let mut result = 0;
        while num > 0 {
            result += (num % 10) as usize;
            num /= 10;
        }

        result
    }
}

// impl Solution {
//     pub fn maximum_sum(nums: Vec<i32>) -> i32 {
//         let mut summed: Vec<(i32, i32)> = nums
//             .iter()
//             .map(|&num| (Self::sum_digits(num), num))
//             .collect();

//         summed.sort_unstable_by(|a, b| b.0.cmp(&a.0).then(b.1.cmp(&a.1)));

//         summed.windows(2).fold(-1, |acc, w| {
//             if w[0].0 == w[1].0 {
//                 acc.max(w[0].1 + w[1].1)
//             } else {
//                 acc
//             }
//         })
//     }

//     #[inline(always)]
//     fn sum_digits(num: i32) -> i32 {
//         let mut num = num;
//         let mut result = 0;
//         while num > 0 {
//             result += num % 10;
//             num /= 10;
//         }

//         result
//     }
// }

#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case([18, 43, 36, 13, 7].to_vec(), 54)]
    #[case([10, 12, 19, 14].to_vec(), -1)]
    fn test(#[case] nums: Vec<i32>, #[case] expected: i32) {
        let result = Solution::maximum_sum(nums);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        ([18, 43, 36, 13, 7].to_vec(), 54),
        ([10, 12, 19, 14].to_vec(), -1),
    ];

    for (nums, expected) in test_cases {
        let result = Solution::maximum_sum(nums.clone());
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            nums, expected, result
        );
        assert_eq!(result, expected);
    }
}
